drivers/w5100/w5100.c

```diff
@@ -247,36 +260,38 @@ static int recv(netdev_t *netdev, void *buf, size_t len, void *info)
         uint16_t rp = raddr(dev, S0_RX_RD0, S0_RX_RD1);
         uint16_t psize = raddr(dev, (S0_RX_BASE + (rp & S0_MASK)),
                                   (S0_RX_BASE + ((rp + 1) & S0_MASK)));
-        n = psize - 2;
+        len = psize - 2;
 
-        DEBUG("[w5100] recv: got packet of %i byte (at 0x%04x)\n", n, (int)rp);
+        DEBUG("[w5100] recv: got packet of %i byte (at 0x%04x)\n", len, (int)rp);
 
         /* read the actual data into the given buffer if wanted */
         if (in_buf != NULL) {
+            /* Is provided buffer big enough? */
+            if (len > max_len) {
+                drop(dev, num, rp, psize);
+                spi_release(dev->p.spi);
+                return -ENOBUFS;
+            }
             uint16_t pos = rp + 2;
-            len = (n <= len) ? n : len;
+
             for (unsigned i = 0; i < len; i++) {
                 in_buf[i] = rreg(dev, (S0_RX_BASE + ((pos++) & S0_MASK)));
             }
 
             DEBUG("[w5100] recv: read %i byte from device (at 0x%04x)\n",
-                  n, (int)rp);
-
-            /* set the new read pointer address */
-            waddr(dev, S0_RX_RD0, S0_RX_RD1, rp += psize);
-            wreg(dev, S0_CR, CR_RECV);
+                  len, (int)rp);
+        }
 
-            /* if RX buffer now empty, clear RECV interrupt flag */
-            if ((num - psize) == 0) {
-                wreg(dev, S0_IR, IR_RECV);
-            }
+        /* if frame received OR drop requested, remove frame from RX buffer */
+        if ((in_buf != NULL) || (len > 0)) {
```

Comment:
I think this must be `&&`?

Reply:
(when I applied this it worked)

Reply:
This is strange. The original code did advance the read pointer address after reading the frame, which is now done in `drop()`. When `||` is changed to `&&` above, the read pointer address would no longer be advanced after reading the frame. I will try to investigate this issue

https://github.com/RIOT-OS/RIOT/blob/af627606f08fd9db74b7f8b927e6b84f43fbe562/drivers/w5100/w5100.c#L254-L273

Reply:
OK, it should have checked for `max_len > 0`, as `len` is computed within this function and always `>0`, unless an empty frame was received.

Also, checking for `max_len > 0` should be sufficient. The cases we have are:

| Case     | `buf`     | `max_len` | Call `drop()`? |
|----------|-----------|-----------|----------------|
| get size | `NULL`    | `0`       | No             |
| recvieve | `!= NULL` | `> 0`     | Yes            |
| drop     | `NULL`    | `> 0`     | Yes            |